client/my-sites/stats/activity-log-day/index.jsx

```diff
@@ -144,47 +186,76 @@ class ActivityLogDay extends Component {
 			requestRestore,
 			rewindConfirmDialog,
 			siteId,
+			tsEndOfSiteDay,
 		} = this.props;
 
-		const hasLogs = ! isEmpty( logs );
+		const hasLogs = ! isEmpty( logs ),
+			rewindHere = this.state.rewindHere,
+			dayExpanded = this.state.dayExpanded ? true : rewindHere;
+
+		let hasConfirmDialog = false;
+
+		const items =
+			hasLogs &&
+			flatMap( logs, log => {
+				const restoreThis = log.activityId === requestedRestoreActivityId;
+				if ( ! hasConfirmDialog ) {
+					hasConfirmDialog =
+						restoreThis &&
+						( tsEndOfSiteDay - DAY_IN_MILLISECONDS <= log.activityTs &&
+							log.activityTs <= tsEndOfSiteDay );
+				}
+				return [
+					restoreThis && rewindConfirmDialog,
+					<ActivityLogItem
+						applySiteOffset={ applySiteOffset }
+						disableRestore={ disableRestore }
+						hideRestore={ hideRestore }
+						key={ log.activityId }
+						log={ log }
+						requestRestore={ requestRestore }
+						siteId={ siteId }
+					/>,
+				];
+			} );
```

Comment:
I'm curious about our use of `let hasConfirmDisalog` here and whether we could get away with more basic assignment as a `const`. what is our purpose of this code? in many cases when we are iterating over a list of things and carrying forth some value what we have is a _reduce_ operation over the list. we supply a function which takes the previous return value and gets the next value and that loops until the end. in this case, we might rewrite this as so…

```js
const [ activityItems, hasConfirmDialog ] = logs.reduce( ( [ items, hasConfirmDialog ], log ) => {
	const restoreThis = log.activityId === requestedRestoreActivityId;

	return [
		[
			...items,
			restoreThis && rewindConfirmDialog,
			<ActivityLogItem … />
		],
		hasConfirmDialog || ( restoreThis && ( … ) )
	];
}, [ [], false ] );
```

We could also use an object as the _accumulator_, the big that the reducer returns and passes along to the next iteration. With this pattern we're never mutating any variables and we can focus on moving from one step to the next instead of thinking more globally (considering `hasConfirmDialog` here).

Reply:
Actually the more I look at it, why is `hasConfirmDialog` inside of this iteration anyway? It doesn't appear to be related to building the items at all and so maybe we could move it out into another computation.

```js
const hasConfirmDialog = logs.some( ( { activityId, activityTs } ) => (
	restoreThis &&
	(
		tsEndOfSiteDay - DAY_IN_MILLISECONDS <= activityTs &&
		activityTs <= tsEndOfSiteDay
	)
) );

const items = flatMap( logs, log => [
	log.activityId === requestedRestoreActivityId && rewindConfirmDialog,
	<ActivityLogItem
		…
	/>
] );
```

Reply:
I didn't want to create an additional loop for this. I know it's not related, but given that we might have a lot of items in a day, I didn't want to go over the same loop again.
I'm also aware it will be more descriptive, but the only reason driving my decision is to make as few calculations as possible.
If you consider that being more descriptive outweighs running the same loop again then I'll make the change.

Reply:
I'd say let's generally be careful in assuming that something will be a problem. We might have up to a thousand activities here, right? So iterating over a thousand things is normal - could be problematic, may not be. Tens of thousands or millions for sure, hundreds probably not.

Maintainability is far more important to me and the general saying is that we spend 80% of our time _reading_ code; therefore, I think it's easy to justify writing the code _first_ in a way that makes it clear what we want or intend on accomplishing and waiting to see if performance is a problem before obscuring the code for the name of speed.

We can note too that the first `.some()` loop should be quite fast and on average will only iterate through half the list (assuming that we have at least one restore point per day and that restores are equally likely to occur at any time). Probably this loop will terminate early on in most iterations because we'll find the restore point we're looking for early in the list.

> If you consider that being more descriptive outweighs running the same loop again then I'll make the change.

So…definitely! :smile:

Another way to think about this is that if we suppose performance _does_ become an issue, how easily will the next person be able to come in, figure out what is supposed to happen, and then fix the issue? We have a number of optimizations available to us, but they are only possible if we can make specific assumptions about the code and its purpose

Reply:
Updated in https://github.com/Automattic/wp-calypso/pull/18722/commits/e09f5df96dc070560ec5b83ce86a5ab3fbfe0172
Thanks for the review 👍